MyFirstSlideshow/app/src/main/java/com/charles/myfirstslideshow/MainActivity.java

```diff
@@ -10,12 +26,141 @@
  *  This code is for demonstration purposes only. It should not be assumed to represent best
  *  practices or patterns employed at Yoti.
  */
+public class MainActivity extends AppCompatActivity implements Serializable {
+
+    // Images under Creative Commons
+    // Images attributed to (in order) https://www.flickr.com/photos/_torne/
+    // https://www.flickr.com/photos/chrisyarzab/
+    // https://www.reddit.com/user/lalien42/
+    // http://www.kapstadt.de/
+
+    public static List images = Arrays.asList(new String[]{"https://c1.staticflickr.com/6/5615/15570202337_0e64f5046e_k.jpg",
+            "https://c1.staticflickr.com/4/3169/2846544061_cb7c04b46f_b.jpg",
+            "https://i.redd.it/d8q1wkgu1awy.jpg",
+            "http://www.kapstadt.de/webcam.jpg"});
 
-public class MainActivity extends AppCompatActivity {
+    static int i = 0;
+    static int j = 0;
+
+    volatile WeakReference<ImageView> image1;
+    volatile WeakReference<ImageView> image2;
+
+    static View[] dots1;
+    static View[] dots2;
 
     @Override
     protected void onCreate(Bundle savedInstanceState) {
-        super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
+
+        if (savedInstanceState == null) {
+            dots1 = new View[4];
+            dots1[0] = getWindow().findViewById(R.id.dots1dot1);
+            dots1[1] = getWindow().findViewById(R.id.dots1dot2);
+            dots1[2] = getWindow().findViewById(R.id.dots1dot3);
+            dots1[3] = getWindow().findViewById(R.id.dots1dot4);
+            dots2 = new View[4];
+            dots2[0] = getWindow().findViewById(R.id.dots2dot1);
+            dots2[1] = getWindow().findViewById(R.id.dots2dot2);
+            dots2[2] = getWindow().findViewById(R.id.dots2dot3);
+            dots2[3] = getWindow().findViewById(R.id.dots2dot4);
+        }
+
+        super.onCreate(savedInstanceState);
+
+        image1 = new WeakReference<ImageView>((ImageView) getWindow().findViewById(R.id.image1));
+        image2 = new WeakReference<ImageView>((ImageView) getWindow().findViewById(R.id.image2));
+    }
+
+    public synchronized void onButtonOneClicked(View button) {
```

Comment:
mainactivity ; 74 this method would be background theard.Because we are making an api call. Android doesn't support network call in main thread which leads to crash giving anr

Reply:
MainActivity : 91 - why using getWindow() use just findViewById.... and also you can create the local variable and then use it.  for readability.